#### Converting the table from the Wikipedia page in a Pandas dataframe.

In [39]:
import pandas as pd
link = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
table = pd.read_html(link)[0]

In [40]:
table.head(15)

Postcode           Borough     Neighbourhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A  Downtown Toronto      Queen's Park
8       M8A      Not assigned      Not assigned
9       M9A      Queen's Park      Not assigned
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
12      M2B      Not assigned      Not assigned
13      M3B        North York   Don Mills North
14      M4B         East York  Woodbine Gardens

In [41]:
no_borough = table[table["Borough"]=="Not assigned"]
table.drop(no_borough.index,inplace = True)
table.reset_index(drop=True)

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M6A        North York   
5        M7A  Downtown Toronto   
6        M9A      Queen's Park   
7        M1B       Scarborough   
8        M1B       Scarborough   
9        M3B        North York   
10       M4B         East York   
11       M4B         East York   
12       M5B  Downtown Toronto   
13       M5B  Downtown Toronto   
14       M6B        North York   
15       M9B         Etobicoke   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M1C       Scarborough   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M3C        North York   
24       M3C        North York   
25       M4C         East York   
26       M5C  Downtown Toronto   
27       M6C              York   
28       M9C         Etobicoke   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M1E       Scarborough   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M4E      East Toronto   
36       M5E  Downtown Toronto   
37       M6E              York   
38       M1G       Scarborough   
39       M4G         East York   
40       M5G  Downtown Toronto   
41       M6G  Downtown Toronto   
42       M1H       Scarborough   
43       M2H        North York   
44       M3H        North York   
45       M3H        North York   
46       M3H        North York   
47       M4H         East York   
48       M5H  Downtown Toronto   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M6H      West Toronto   
52       M6H      West Toronto   
53       M1J       Scarborough   
54       M2J        North York   
55       M2J        North York   
56       M2J        North York   
57       M3J        North York   
58       M3J        North York   
59       M4J         East York   
60       M5J  Downtown Toronto   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M6J      West Toronto   
64       M6J      West Toronto   
65       M1K       Scarborough   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M2K        North York   
69       M3K        North York   
70       M3K        North York   
71       M4K      East Toronto   
72       M4K      East Toronto   
73       M5K  Downtown Toronto   
74       M5K  Downtown Toronto   
75       M6K      West Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M1L       Scarborough   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M2L        North York   
82       M2L        North York   
83       M3L        North York   
84       M4L      East Toronto   
85       M4L      East Toronto   
86       M5L  Downtown Toronto   
87       M5L  Downtown Toronto   
88       M6L        North York   
89       M6L        North York   
90       M6L        North York   
91       M9L        North York   
92       M1M       Scarborough   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M2M        North York   
96       M2M        North York   
97       M3M        North York   
98       M4M      East Toronto   
99       M5M        North York   
100      M5M        North York   
101      M6M              York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M9M        North York   
106      M9M        North York   
107      M1N       Scarborough   
108      M1N       Scarborough   
109      M2N        North York   
110      M3N        North York   
111      M4N   Central Toronto   
112      M5N   Central Toronto   
113      M6N              York   
114      M6N              York   
115      M9N              York   
116      M1P       Scarbor

In [42]:
s = table.shape
for i in range(0,s[0]):
    if table["Neighbourhood"].iloc[i]=="Not assigned":
        table.replace("Not assigned",table["Borough"].iloc[i],inplace = True)
        

In [43]:
table.reset_index(drop=True,inplace = True)

In [44]:
table.head(20)

Postcode           Borough     Neighbourhood
0       M3A        North York         Parkwoods
1       M4A        North York  Victoria Village
2       M5A  Downtown Toronto      Harbourfront
3       M6A        North York  Lawrence Heights
4       M6A        North York    Lawrence Manor
5       M7A  Downtown Toronto      Queen's Park
6       M9A      Queen's Park      Queen's Park
7       M1B       Scarborough             Rouge
8       M1B       Scarborough           Malvern
9       M3B        North York   Don Mills North
10      M4B         East York  Woodbine Gardens
11      M4B         East York     Parkview Hill
12      M5B  Downtown Toronto           Ryerson
13      M5B  Downtown Toronto   Garden District
14      M6B        North York         Glencairn
15      M9B         Etobicoke        Cloverdale
16      M9B         Etobicoke         Islington
17      M9B         Etobicoke      Martin Grove
18      M9B         Etobicoke  Princess Gardens
19      M9B         Etobicoke   West Deane Park

In [45]:
table.shape

(210, 3)

In [46]:
table_new = table.groupby(["Postcode","Borough"],axis = 0)["Neighbourhood"].apply(lambda x: ','.join(x))
table_new  = pd.DataFrame(table_new)
table_new.reset_index(drop=False,inplace = True)

In [47]:
table_new.head(60)

Postcode           Borough  \
0       M1B       Scarborough   
1       M1C       Scarborough   
2       M1E       Scarborough   
3       M1G       Scarborough   
4       M1H       Scarborough   
5       M1J       Scarborough   
6       M1K       Scarborough   
7       M1L       Scarborough   
8       M1M       Scarborough   
9       M1N       Scarborough   
10      M1P       Scarborough   
11      M1R       Scarborough   
12      M1S       Scarborough   
13      M1T       Scarborough   
14      M1V       Scarborough   
15      M1W       Scarborough   
16      M1X       Scarborough   
17      M2H        North York   
18      M2J        North York   
19      M2K        North York   
20      M2L        North York   
21      M2M        North York   
22      M2N        North York   
23      M2P        North York   
24      M2R        North York   
25      M3A        North York   
26      M3B        North York   
27      M3C        North York   
28      M3H        North York   
29      M3J        North York   
30      M3K        North York   
31      M3L        North York   
32      M3M        North York   
33      M3N        North York   
34      M4A        North York   
35      M4B         East York   
36      M4C         East York   
37      M4E      East Toronto   
38      M4G         East York   
39      M4H         East York   
40      M4J         East York   
41      M4K      East Toronto   
42      M4L      East Toronto   
43      M4M      East Toronto   
44      M4N   Central Toronto   
45      M4P   Central Toronto   
46      M4R   Central Toronto   
47      M4S   Central Toronto   
48      M4T   Central Toronto   
49      M4V   Central Toronto   
50      M4W  Downtown Toronto   
51      M4X  Downtown Toronto   
52      M4Y  Downtown Toronto   
53      M5A  Downtown Toronto   
54      M5B  Downtown Toronto   
55      M5C  Downtown Toronto   
56      M5E  Downtown Toronto   
57      M5G  Downtown Toronto   
58      M5H  Downtown Toronto   
59      M5J  Downtown Toronto   

                                        Neighbourhood  
0                                       Rouge,Malvern  
1                Highland Creek,Rouge Hill,Port Union  
2                     Guildwood,Morningside,West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                                 Scarborough Village  
6           East Birchmount Park,Ionview,Kennedy Park  
7                       Clairlea,Golden Mile,Oakridge  
8       Cliffcrest,Cliffside,Scarborough Village West  
9                          Birch Cliff,Cliffside West  
10  Dorset Park,Scarborough Town Centre,Wexford He...  
11                                   Maryvale,Wexford  
12                                          Agincourt  
13              Clarks Corners,Sullivan,Tam O'Shanter  
14  Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                                    L'Amoreaux West  
16                                        Upper Rouge  
17                                  Hillcrest Village  
18                         Fairview,Henry Farm,Oriole  
19                                    Bayview Village  
20                            Silver Hills,York Mills  
21                             Newtonbrook,Willowdale  
22                                   Willowdale South  
23                                    York Mills West  
24                                    Willowdale West  
25                                          Parkwoods  
26                                    Don Mills North  
27                    Flemingdon Park,Don Mills South  
28      Bathurst Manor,Downsview North,Wilson Heights  
29                     Northwood Park,York University  
30                         CFB Toronto,Downsview East  
31                                     Downsview West  
32                                  Downsview Central  
33                                Downsview Northwest  
34                           

#### Importing required libraries.

In [48]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium

#### Reading the location CSV file-

In [49]:
loc_data = pd.read_csv("Geospatial_Coordinates.csv")

In [50]:
loc_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

#### As we can see, both of the daraframes has the same number of rows,i.e. 103.

#### Renaming the column "Postal code" as "Postcode" so that we can merge the dataframes.

In [51]:
loc_data.rename(columns={"Postal Code":"Postcode"},inplace = True)

In [52]:
loc_data.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [53]:
loc_data.shape

(103, 3)

#### Merging both the dataframes to get the desired dataframe which has "Postcode","Borough","Neighbourhood","Latitude" and "Longitude".

In [54]:
table_with_location = pd.merge(table_new,loc_data,how = "outer")
table_with_location.rename(columns = {"Postcode":"PostalCode"},inplace = True)
table_with_location.head()

PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### We need the location data of Toronto to segment and cluster the neighbourhood.

In [55]:
table_borough = table_with_location[["Borough"]]
toronto_table = table_with_location[table_borough.apply(lambda row: row.astype(str).str.contains('Toronto').any(), axis=1)]
toronto_table

PostalCode           Borough  \
37        M4E      East Toronto   
41        M4K      East Toronto   
42        M4L      East Toronto   
43        M4M      East Toronto   
44        M4N   Central Toronto   
45        M4P   Central Toronto   
46        M4R   Central Toronto   
47        M4S   Central Toronto   
48        M4T   Central Toronto   
49        M4V   Central Toronto   
50        M4W  Downtown Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
63        M5N   Central Toronto   
64        M5P   Central Toronto   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
68        M5V  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
75        M6G  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   
85        M7A  Downtown Toronto   
87        M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                        The Danforth West,Riverdale  43.679557 -79.352188  
42                      The Beaches West,India Bazaar  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                         Moore Park,Summerhill East  43.689574 -79.383160  
49  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                         Cabbagetown,St. James Town  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                                       Harbourfront  43.654260 -79.360636  
54                            Ryerson,Garden District  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                             Adelaide,King,Richmond  43.650571 -79.384568  
59    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
60            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
61                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
65                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
66                      Harbord,University of Toronto  43.662696 -79.400049  
67            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
68  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
69                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
70              First Canadian Place,Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                        Dovercourt Village,Dufferin  43.669005 -79.442259  
77                          

#### Getting the Latitude and Longitude of Toronto, Canada - 

In [57]:
address = 'Toronto, CANADA'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Creating a map of Toronto using Folium library

In [59]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

#### Superimposing the neighbourhoods obtained from the dataframe <b>"toronto_table"</b> on the <b>"map_toronto"</b>-

In [60]:
for lat, lng, borough, neighbourhood in zip(toronto_table['Latitude'], toronto_table['Longitude'], toronto_table['Borough'], toronto_table['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto